### PR TITLE
Implement batch training for new open ff codebase

### PR DESCRIPTION
**Pull request by chuantian**

### CODE REVIEW

File: fe/parameterize_batch.py [hunk 1]
```diff
@@ -0,0 +1,523 @@
...
+                lambda_schedule,
+                lambda_idxs,
+                precision
+            )
+
+            epoch_train_params = get_params(opt_state)
+
+            epoch_train_ff_params = copy.deepcopy(open_ff)
+            epoch_train_ff_params.params = epoch_train_params[len(host_system.params):]
+            fname = "epoch_"+str(epoch)+"_dG_"+str(true_dG)+"_params"
```
Comment (proteneer) [inline fe/parameterize_batch.py]: why do we put the true_dG in the filename?

File: fe/parameterize_batch.py [hunk 2]
```diff
@@ -0,0 +1,523 @@
...
+                du_dls = pc.recv()
+                all_du_dls.append(du_dls)
+
+            all_du_dls = np.array(all_du_dls)
+            loss_grad_fn = jax.grad(error_fn, argnums=(0,)) 
+
+            error = error_fn(all_du_dls, T, lambda_schedule, true_dG)
+
+            print("---EPOCH", epoch, "---Train---LOSS", error)
+            with open(args.out_dir+"/loss.txt", "a+") as f:
```
Comment (proteneer) [inline fe/parameterize_batch.py]: always do os.path.join(args.out_dir, "loss.txt")

Comment (proteneer) [reply fe/parameterize_batch.py]: (Also do this for the earlier check to see if the file exists)

File: fe/parameterize_batch.py [hunk 3]
```diff
@@ -0,0 +1,523 @@
...
+
+            guest_mol = copy.deepcopy(data[0][0])
+            true_dG = data[0][1] # in kJ/mol and positive sign 
+
+            init_conf = guest_mol.GetConformer(0)
+            init_conf = np.array(init_conf.GetPositions(), dtype=np.float64)
+            init_conf = init_conf/10 # convert to md_units
+            conf_com = com(init_conf)
+
+            init_mol = Chem.Mol(guest_mol)
```
Comment (proteneer) [inline fe/parameterize_batch.py]: Now I see why you need to use the deepcopy!

File: fe/parameterize_batch.py [hunk 4]
```diff
@@ -0,0 +1,523 @@
...
+            amber_ff = app.ForceField('amber99sbildn.xml', 'amber99_obc.xml')
+            host_system = amber_ff.createSystem(host_pdb.topology,
+                nonbondedMethod=app.NoCutoff,
+                constraints=None,
+                rigidWater=False)
+
+            host_system = openmm_converter.deserialize_system(host_system)
+            num_host_atoms = len(host_system.masses)
+
+            # open_ff = forcefield.Forcefield("ff/smirnoff_1.1.0.py")
```
Comment (proteneer) [inline fe/parameterize_batch.py]: we can probably remove this comment

File: fe/parameterize_batch.py [hunk 5]
```diff
@@ -0,0 +1,523 @@
...
+                        print("derivs", g_idx, '\t group', gp, '\t', g, '\t adjusted to', g*pf, '\t old val', sim.system.params[g_idx])
+                else:
+                    filtered_grad.append(0)
+
+            filtered_grad = np.array(filtered_grad)
+            opt_state = opt_update(epoch, filtered_grad, opt_state)
+
+        for data in test.iterbatches(1):
+
+            guest_mol = copy.deepcopy(data[0][0])
```
Comment (proteneer) [inline fe/parameterize_batch.py]: The test set has a *lot* of duplicate code shared with the training set - can you find a way to refactor them out? If we change something in the training code we'd also have to remember to change it exactly the same way in the test code. 

For example
- the host_pdb code is identical, and can be moved entirely out of the training loop
- lambda schedules etc are the same.

File: fe/parameterize_batch.py [hunk 6]
```diff
@@ -0,0 +1,523 @@
...
+
+    for ligand_sdf_file in args.ligand_sdf:
+        for rd_mol in Chem.SDMolSupplier(ligand_sdf_file):
+            dG_value = rd_mol.GetProp("_dG")
+            ligand_list.append((rd_mol, float(dG_value)))
+
+    ligand_dataset = dataset.Dataset(ligand_list)
+    train, test = ligand_dataset.split(0.7)
+
+    if os.path.exists(args.out_dir+"/loss.txt"):
```
Comment (proteneer) [inline fe/parameterize_batch.py]: what I've been doing is just pipe runs to to a file then grep for LOSS afterwards

File: fe/parameterize_batch.py [hunk 7]
```diff
@@ -0,0 +1,523 @@
...
+                p = Process(target=sim.run_forward_and_backward, args=arg)
+                p.daemon = True
+                processes.append(p)
+                p.start()
+
+            all_du_dls = []
+            for pc in parent_conns:
+                du_dls = pc.recv()
+                all_du_dls.append(du_dls)
+
```
Comment (proteneer) [inline fe/parameterize_batch.py]: you need to close parent conns with pc.close() here else it will leak processes (run nvidia-smi and you should see memory usage start to creep up). You will also run out of fds on your os. 

I think you will also have to wrap the simulation.py code with:
``` python
simulation.py: run_forward_and_backward()

try:
        du_dl_adjoints = pipe.recv()
        stepper.set_du_dl_adjoint(du_dl_adjoints)
        ctxt.set_x_t_adjoint(np.zeros_like(x0))
        start = time.time()
        ctxt.backward_mode()
        print("bkwd run time", time.time() - start)
        dL_dp = ctxt.get_param_adjoint_accum()
        pipe.send(dL_dp)
except EOFError:
        pass
pipe.close()
return
```

File: fe/parameterize_batch.py (hunk 1 above)
Comment (chuantian) [reply fe/parameterize_batch.py]: I use ligand_id just for now, but it won't associate with one particular ligand every epoch since we do shuffle.

File: fe/parameterize_batch.py (hunk 6 above)
Comment (chuantian) [reply fe/parameterize_batch.py]: Yes, that's okay too. I save this just for monitoring the loss more easily.

File: fe/parameterize_batch.py [hunk 8]
```diff
@@ -0,0 +1,523 @@
...
+                parent_conn, child_conn = Pipe()
+                parent_conns.append(parent_conn)
+                # writer can be None if we don't care about vis
+                all_args.append([x0, v0, conf_idx % num_gpus, writer, child_conn])
+
+            processes = []
+
+            for arg in all_args:
+                p = Process(target=sim.run_forward_and_backward, args=arg)
```
Comment (proteneer) [inline fe/parameterize_batch.py]: Process() called once here

File: fe/parameterize_batch.py [hunk 9]
```diff
@@ -0,0 +1,523 @@
...
```
Comment (proteneer) [inline fe/parameterize_batch.py]: Process() called second time here

File: fe/parameterize_batch.py [hunk 10]
```diff
@@ -0,0 +1,487 @@
...
+
+            all_du_dls = []
+            for pc in parent_conns:
+                du_dls = pc.recv()
+                all_du_dls.append(du_dls)           
+                pc.close()
+
+            # terminate all the processes
+            for p in processes:
+                p.terminate()
```
Comment (chuantian) [inline fe/parameterize_batch.py]: I have to do p.terminate in the test batch instead of p.join because the process doesn't join properly.

Comment (proteneer) [reply fe/parameterize_batch.py]: hm why doesn't it join properly? that's a little concerning

**Comment by proteneer** [comment]
Thanks for the excellent contribution. Merging
